cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -779,6 +779,11 @@ ParquetFileFragment::ParquetFileFragment(FileSource source,
       parquet_format_(checked_cast<ParquetFileFormat&>(*format_)),
       row_groups_(std::move(row_groups)) {}
 
+const std::shared_ptr<parquet::FileMetaData>& ParquetFileFragment::metadata() {
+  auto lock = physical_schema_mutex_.Lock();
+  return metadata_;
```

Comment:
This is returning a reference to an internal member, but the dereferencing won't be protected by the lock.
Should we switch to return a `std::shared_ptr` instead?

Reply:
Yikes, yes, you are right.  I've updated this to return by value instead of by reference.